subsys/debug/gpio_frontend/gpio_frontend.c

```diff
@@ -0,0 +1,677 @@
+		"STR R2, [R1, %[set_address_offset]]	\n"
+	: /* Output */
+	: /* Input */
+		[str]"r"((u32_t)str),
+		[src_level]"r"(log_msg_ids_bytes.raw),
+		[arg0]"r"((u32_t)arg0),
+		[arg1]"r"((u32_t)arg1),
+		[arg2]"r"((u32_t)arg2),
+		[base_address]"r"(BASE_ADDRESS),
+		[mask_of_clk_pin]"r"(MASK_OF_CLK_PIN),
+		[mask_of_log_pins]"r"(MASK_OF_LOG_PINS),
+		[lsb_pin_number]"i"(CONFIG_LOG_FRONTEND_LSB_PIN_NUMBER),
+		[clear_address_offset]"i"(CLEAR_ADDRESS_OFFSET),
+		[set_address_offset]"i"(SET_ADDRESS_OFFSET)
+		: "r0", "r1", "r2", "r3"
+	);
+
+	irq_unlock(key);
+}
+
+void log_frontend_n(const char *str,
+		    log_arg_t *args,
+		    u32_t narg,
+		    struct log_msg_ids src_level)
+{
+	union log_msg_ids_convert log_msg_ids_bytes;
+
+	log_msg_ids_bytes.ids = src_level;
+
+	unsigned int key = irq_lock();
+
+	__asm volatile(
+		/* str - first byte */
+		"MOV R0, %[str]				\n"
+		"MOV R1, %[base_address]		\n"
+		"MOV R2, %[mask_of_clk_pin]		\n"
+		"MOV R3, %[mask_of_log_pins]		\n"
+		"BFI R2, R0, %[lsb_pin_number], #8	\n"
+		"STR R3, [R1, %[clear_address_offset]]	\n"
+		"STR R2, [R1, %[set_address_offset]]	\n"
+		/* str - second byte */
+		"LSR R0, #8				\n"
+		"BFI R2, R0, %[lsb_pin_number], #8	\n"
+		"STR R3, [R1, %[clear_address_offset]]	\n"
+		"STR R2, [R1, %[set_address_offset]]	\n"
+		/* str - third byte */
+		"LSR R0, #8				\n"
+		"BFI R2, R0, %[lsb_pin_number], #8	\n"
+		"STR R3, [R1, %[clear_address_offset]]	\n"
+		"STR R2, [R1, %[set_address_offset]]	\n"
+		/* str - fourth byte */
+		"LSR R0, #8				\n"
+		"BFI R2, R0, %[lsb_pin_number], #8	\n"
+		"STR R3, [R1, %[clear_address_offset]]	\n"
+		"STR R2, [R1, %[set_address_offset]]	\n"
+	: /* Output */
+	: /* Input */
+		[str]"r"((u32_t)str),
+		[base_address]"r"(BASE_ADDRESS),
+		[mask_of_clk_pin]"r"(MASK_OF_CLK_PIN),
+		[mask_of_log_pins]"r"(MASK_OF_LOG_PINS),
+		[lsb_pin_number]"i"(CONFIG_LOG_FRONTEND_LSB_PIN_NUMBER),
+		[clear_address_offset]"i"(CLEAR_ADDRESS_OFFSET),
+		[set_address_offset]"i"(SET_ADDRESS_OFFSET)
+		: "r0", "r1", "r2", "r3"
+	);
+	for (u32_t i = 0; i < narg; i++) {
+		__asm volatile(
+			/* args[i] - first byte */
+			"MOV R0, %[arg_i]			\n"
+			"MOV R1, %[base_address]		\n"
+			"MOV R2, %[mask_of_clk_pin]		\n"
+			"MOV R3, %[mask_of_log_pins]		\n"
+			"BFI R2, R0, %[lsb_pin_number], #8	\n"
+			"STR R3, [R1, %[clear_address_offset]]	\n"
+			"STR R2, [R1, %[set_address_offset]]	\n"
+			/* args[i] - second byte */
+			"LSR R0, #8				\n"
+			"BFI R2, R0, %[lsb_pin_number], #8	\n"
+			"STR R3, [R1, %[clear_address_offset]]	\n"
+			"STR R2, [R1, %[set_address_offset]]	\n"
+			/* args[i] - third byte */
+			"LSR R0, #8				\n"
+			"BFI R2, R0, %[lsb_pin_number], #8	\n"
+			"STR R3, [R1, %[clear_address_offset]]	\n"
+			"STR R2, [R1, %[set_address_offset]]	\n"
+			/* args[i] - fourth byte */
+			"LSR R0, #8				\n"
+			"BFI R2, R0, %[lsb_pin_number], #8	\n"
+			"STR R3, [R1, %[clear_address_offset]]	\n"
+			"STR R2, [R1, %[set_address_offset]]	\n"
```

Comment:
A description of the algorithm, preferably with an accompanying C equivalent is needed
for this code to be understandable.